std/algorithm/comparison.d

```diff
@@ -892,116 +892,353 @@ nothrow pure @safe unittest
+            import std.meta : staticMap, staticIndexOf;
+
+            // get lengths
+            alias hasLengthBits = staticMap!(hasLength, Rs);
+            enum indexOfFirstLength = staticIndexOf!(true, hasLengthBits);
+            enum someHasLength = indexOfFirstLength != -1;
 
-            //Lengths are the same, so we need to do an actual comparison
-            //Good news is we can squeeze out a bit of performance by not checking if r2 is empty
-            for (; !r1.empty; r1.popFront(), r2.popFront())
+            static if (someHasLength && // if one range has a (finite) length and
+                       anySatisfy!(isInfinite, Rs)) // another is infinite
             {
-                if (!binaryFun!(pred)(r1.front, r2.front)) return false;
+                return false;   // they can't be equal
             }
-            return true;
-        }
-        else
-        {
-            //Generic case, we have to walk both ranges making sure neither is empty
-            for (; !r1.empty; r1.popFront(), r2.popFront())
+            else
             {
-                if (r2.empty) return false;
-                if (!binaryFun!(pred)(r1.front, r2.front)) return false;
+                enum someHasIndexingAndLength = anySatisfy!(hasIndexingAndLength, Rs);
+
+                static if (someHasLength)
+                {
+                    /* if any `rs` `hasLength` then `primaryRangeIndex` will index the
+                     * first of them */
+                    enum primaryRangeIndex = indexOfFirstLength;
+                    const length = rs[indexOfFirstLength].length;
+                }
+                else
+                {
+                    enum primaryRangeIndex = 0; // just pick first
+                }
+                alias r_ = rs[primaryRangeIndex]; // shorthand
+
+                // check lengths
+                static if (someHasLength)
+                {
+                    static foreach (i, r; rs)
+                    {
+                        static if (i != indexOfFirstLength)
+                        {
+                            static if (hasLength!(typeof(r)))
+                            {
+                                if (length != r.length) { return false; }
+                            }
+                        }
+                    }
+                }
+
+                // check equal contents
+                static if (someHasIndexingAndLength) // if some `Rs` needs a element offset counter
+                {
+                    size_t ei = 0; // element offset counter
+                }
+                while (true) // for each element in primary range `r`
+                {
+                    static if (hasIndexingAndLength!(typeof(r_)))
+                    {
+                        if (ei == r_.length) { break; }
+                    }
+                    else
+                    {
+                        if (r_.empty) { break; }
+                    }
```

Comment:
Can't we do the check for length directly here?

Reply:
I don't follow you. Can you elaborate? And write the replacement-code in a comment?

Reply:
The idea was instead of checking the length separately, to keep track of the number of iterated elements and thus being able to check for the length directly at the end of the iteration (if `length` is defined).

I will try to produce a diff to your PR later today.